src/transformers/models/gptj/modeling_gptj.py

```diff
@@ -0,0 +1,1046 @@
+
+        self.register_buffer("masked_bias", torch.tensor(-1e9))
+
+        self.attn_pdrop = nn.Dropout(config.attn_pdrop)
+        self.resid_pdrop = nn.Dropout(config.resid_pdrop)
+
+        self.embed_dim = config.n_embd
+        self.num_attention_heads = config.num_attention_heads
+        self.head_dim = self.embed_dim // self.num_attention_heads
+        if config.jax:
+            self.register_buffer("scale_attn", torch.sqrt(torch.tensor(self.head_dim)))
+        else:
+            self.scale_attn = None
+        if self.head_dim * self.num_attention_heads != self.embed_dim:
+            raise ValueError(
+                f"embed_dim must be divisible by num_attention_heads (got `embed_dim`: {self.embed_dim} and `num_attention_heads`: {self.num_attention_heads})."
+            )
+
+        self.k_proj = nn.Linear(self.embed_dim, self.embed_dim, bias=False)
+        self.v_proj = nn.Linear(self.embed_dim, self.embed_dim, bias=False)
+        self.q_proj = nn.Linear(self.embed_dim, self.embed_dim, bias=False)
+        self.out_proj = nn.Linear(self.embed_dim, self.embed_dim, bias=not config.jax)
+        self.rotary = config.rotary
+        self.rotary_dim = None
+        if config.rotary_dim is not None:
+            self.rotary_dim = config.rotary_dim
+
+    def forward(
+        self,
+        hidden_states,
+        attention_mask=None,
+        layer_past=None,
+        head_mask=None,
+        use_cache=False,
+        output_attentions=False,
+    ):
+
+        query = self.q_proj(hidden_states)
+        key = self.k_proj(hidden_states)
+        value = self.v_proj(hidden_states)
+
+        query = self._split_heads(query, self.num_attention_heads, self.head_dim, self.rotary)
+        key = self._split_heads(key, self.num_attention_heads, self.head_dim, self.rotary)
+        value = self._split_heads(value, self.num_attention_heads, self.head_dim, False)
+
+        if self.rotary:
+            seq_len = key.shape[1]
+            offset = 0
+            if layer_past is not None:
+                offset = layer_past[0].shape[-2]
+                seq_len += offset
+            if self.rotary_dim is not None:
+                k_rot = key[:, :, :, :self.rotary_dim]
+                k_pass = key[:, :, :, self.rotary_dim:]
+
+                q_rot = query[:, :, :, :self.rotary_dim]
+                q_pass = query[:, :, :, self.rotary_dim:]
+
+                sincos = fixed_pos_embedding(k_rot, 1, seq_len=seq_len)
+                k_rot = apply_rotary_pos_emb(k_rot, sincos, offset=offset)
+                q_rot = apply_rotary_pos_emb(q_rot, sincos, offset=offset)
+
+                key = torch.cat([k_rot, k_pass], dim=-1)
+                query = torch.cat([q_rot, q_pass], dim=-1)
+            else:
+                sincos = fixed_pos_embedding(key, 1, seq_len=seq_len)
+                key = apply_rotary_pos_emb(key, sincos, offset=offset)
+                query = apply_rotary_pos_emb(query, sincos, offset=offset)
```

Comment:
Very cool!

Think would be nice to keep this in a static method, so it could be tested.